src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -227,11 +227,34 @@ double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::Evaluate(
 
 template<typename OutputLayerType, typename InitializationRuleType,
          typename... CustomLayers>
-void FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::Gradient(
-    const arma::mat& parameters,
-    const size_t begin,
-    arma::mat& gradient,
-    const size_t batchSize)
+double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::Evaluate(
+    const arma::mat& parameters, const size_t begin, const size_t batchSize)
+{
+  return Evaluate(parameters, begin, batchSize, true);
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+         typename... CustomLayers>
+template<typename GradType>
+double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::
+EvaluateWithGradient(const arma::mat& parameters, GradType& gradient)
+{
+  double res = 0;
+  for (size_t i = 0; i < predictors.n_cols; ++i)
+    res += EvaluateWithGradient(parameters, i, gradient, 1, false);
+
+  return res;
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+         typename... CustomLayers>
+template<typename GradType>
+double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::
+EvaluateWithGradient(const arma::mat& /* parameters */,
+                     const size_t begin,
+                     GradType& gradient,
+                     const size_t batchSize,
+                     const bool deterministic)
```

Comment:
Do we (or will we) ever call with `deterministic = true`?  I think that when `determinstic = true`, we are just evaluating the network so there is only a need for a forward pass, in which case we'd call `Evaluate()` instead.  So could the parameter be removed?  It's possible there's something I've overlooked here, so let me know if so. :)

Reply:
@rcurtin Even I am not too sure of this, I think @zoq might be able to help with this?

Reply:
Correct, there is no need for ``deterministic`` in ``EvaluateWithGradient``.